### PR TITLE
Handshake response object if handshake failed

### PR DESCRIPTION
**Pull request by a-palagin**
Added handshake response object, so if error occurs on websocket handshake stage, server's response can be reachable for further decisions

### CODE REVIEW

File: websocket/src/handshake.cpp [hunk 1]
```diff
@@ -170,105 +171,100 @@ Result ReceiveHeaders(WebsocketConnection* conn)
...
     }
 
-    // The response might contain both the headers, but also (if successful) the first batch of data
-    endtag += 4;
-    uint32_t size = conn->m_BufferSize - (endtag - conn->m_Buffer);
-    conn->m_BufferSize = size;
-    memmove(conn->m_Buffer, endtag, size);
-    conn->m_Buffer[size] = 0;
-    conn->m_HasHandshakeData = conn->m_BufferSize != 0 ? 1 : 0;
+    HttpHeader *connection_header, *upgrade_header, *websoket_secret_header;
```
Comment (JCash) [inline websocket/src/handshake.cpp]: typo: websoket_secret_header -> websocket_secret_header

File: websocket/src/websocket.h [hunk 2]
```diff
@@ -78,6 +78,30 @@ namespace dmWebsocket
         uint32_t m_Type:2;
     };
 
+    struct HttpHeader
+    {
+        const char* m_Key;
+        const char* m_Value;
+
+        HttpHeader() = delete;
```
Comment (JCash) [inline websocket/src/websocket.h]: Although some compilers have their default set to C++11, we aren't (yet) officially supporting more than C++98. If the extension compiles stand-alone, for all platforms, it's good.

**Review by JCash — APPROVED** [review]
Looks good.
Some small notes to fix/check before I can merge into master.

File: websocket/src/handshake.cpp (hunk 1 above)
Comment (a-palagin) [reply websocket/src/handshake.cpp]: done

File: websocket/src/websocket.h (hunk 2 above)
Comment (a-palagin) [reply websocket/src/websocket.h]: removed c++11 constructors

File: websocket/src/websocket.cpp [hunk 3]
```diff
@@ -67,6 +67,19 @@ const char* StateToString(State err)
 
 #undef STRING_CASE
 
+int dmStriCmp(const char* s1, const char* s2)
```
Comment (britzl) [inline websocket/src/websocket.cpp:70]: We do have dmStrCaseCmp: https://defold.com/ref/stable/dmStringFunc/#dmStrCaseCmp:s1-s2

Could that be used instead?

**Review by JCash — APPROVED** [review]

File: websocket/src/websocket.cpp (hunk 3 above)
Comment (JCash) [reply websocket/src/websocket.cpp:70]: It could. We'll make a mental note of it and do that in another PR.

**Pull request merged**